File: match_demo.lua
-- Sapphire Lua 5.1 Runtime Header

local Arena = {}
Arena.__index = Arena

function Arena.init()
  local self = setmetatable({}, Arena)
  self.objects = {}
  return self
end

function Arena:register(obj)
  if type(obj) == "table" then
    local meta = getmetatable(obj)
    if meta and meta.__arena and meta.__arena ~= self then
      for i, o in ipairs(meta.__arena.objects) do
        if o == obj then
          table.remove(meta.__arena.objects, i)
          break
        end
      end
    end
    table.insert(self.objects, obj)
    if meta then
      meta.__arena = self
    end
  end
  return obj
end

function Arena:destroy()
  for _, obj in ipairs(self.objects) do
    local meta = getmetatable(obj)
    if meta and meta.__shadow then
      for k in pairs(meta.__shadow) do
        meta.__shadow[k] = nil
      end
    end
  end
  self.objects = {}
end

local _DEFAULT_ARENA = Arena.init()

local _clone_helper

local function _create_proto_object(proto, class_tbl)
  local shadow = {}
  local meta = {}
  meta.__shadow = shadow
  meta.__proto = proto
  meta.__class = class_tbl
  meta.__index = function(tbl, key)
    if key == "__proto__" then return meta.__proto end
    if key == "__shadow__" then return meta.__shadow end
    if shadow[key] ~= nil then return shadow[key] end
    if proto ~= nil then
      local val = proto[key]
      if val ~= nil then
        if type(val) == "table" then
          local cow = _clone_helper(val, nil, nil)
          shadow[key] = cow
          return cow
        end
        return val
      end
    end
    if class_tbl and class_tbl[key] ~= nil then
      return class_tbl[key]
    end
    return nil
  end
  meta.__newindex = function(tbl, key, val)
    if key == "__proto__" or key == "__shadow__" then return end
    if proto ~= nil then
      shadow[key] = val
    else
      rawset(tbl, key, val)
    end
  end
  local obj = setmetatable({}, meta)
  if proto == nil then
    _DEFAULT_ARENA:register(obj)
  end
  return obj
end

_clone_helper = function(obj, init_fn, arena)
  if obj == nil then return nil end
  local clone_obj
  if type(obj) == "table" then
    local meta = getmetatable(obj)
    if obj.clone and type(obj.clone) == "function" then
      clone_obj = obj:clone()
    elseif meta and meta.__class then
      clone_obj = _create_proto_object(obj, meta.__class)
    else
      clone_obj = _create_proto_object(obj, nil)
    end
  else
    clone_obj = obj
  end
  if arena ~= nil and type(clone_obj) == "table" then
    arena:register(clone_obj)
  end
  if init_fn then
    init_fn(clone_obj)
  end
  return clone_obj
end


local Direction = {
  North = 0,
  East = 1,
  South = 2,
  West = 3
}


local function get_direction_name(dir)
  local _match_res_1 = nil
  local _subj_1 = dir
  if _subj_1 == Direction.North then
    _match_res_1 = "North"
  elseif _subj_1 == Direction.East then
    _match_res_1 = "East"
  elseif _subj_1 == Direction.South then
    _match_res_1 = "South"
  else
    _match_res_1 = "West"
  end
  local name = _match_res_1
  return name
end


local function handle_action(code)
  local _match_res_2 = nil
  local _subj_2 = code
  if _subj_2 == 200 then
    _match_res_2 = "OK"
  elseif _subj_2 == 404 then
    _match_res_2 = "Not Found"
  else
    local log_msg = "Unknown status code"
    _match_res_2 = log_msg
  end
  local result = _match_res_2
  return result
end


local function run_side_effects(dir)
  local _subj_3 = dir
  if _subj_3 == Direction.North then
    local step = 1
  else
    local step = 0
  end
end


local function main()
  local dir_name = get_direction_name(Direction.South)
  local status_str = handle_action(404)
  run_side_effects(Direction.East)
  if ((dir_name == "South") and (status_str == "Not Found")) then
    return 0
  end
  return 1
end


main()
main()

File: samples/match_demo.py
# Sapphire Runtime Header
import copy
from enum import Enum, IntEnum

class Arena:
  def __init__(self):
    self.objects = []
  def register(self, obj):
    if hasattr(obj, '__arena__') and obj.__arena__ is not None:
      if obj.__arena__ is not self:
        try:
          obj.__arena__.objects.remove(obj)
        except ValueError:
          pass
    if obj not in self.objects:
      self.objects.append(obj)
    if hasattr(obj, '__setattr__'):
      try:
        object.__setattr__(obj, '__arena__', self)
      except AttributeError:
        pass
    return obj
  def destroy(self):
    for obj in self.objects:
      if hasattr(obj, '__shadow__'):
        obj.__shadow__.clear()
    self.objects.clear()
  def __enter__(self):
    return self
  def __exit__(self, exc_type, exc_val, exc_tb):
    self.destroy()

_DEFAULT_ARENA = Arena()

class SapphireObject:
  def __init__(self, proto=None):
    super().__setattr__('__proto__', proto)
    super().__setattr__('__shadow__', {})
    if proto is None:
      _DEFAULT_ARENA.register(self)

  def clone(self):
    clone_obj = self.__class__(proto=self)
    if hasattr(self, '__arena__') and self.__arena__ is not None:
      self.__arena__.register(clone_obj)
    return clone_obj

  def __getattr__(self, name):
    if name.startswith('__') and name.endswith('__'):
      if name == '__proto__':
        return self.__proto__
      raise AttributeError(f"Attribute '{name}' not found on {self.__class__.__name__}")
    if name in self.__shadow__:
      return self.__shadow__[name]
    if self.__proto__ is not None:
      val = getattr(self.__proto__, name)
      if not isinstance(val, (int, float, bool, str, type(None))):
        if hasattr(val, 'clone'):
          cow_val = val.clone()
        else:
          cow_val = copy.deepcopy(val)
        self.__shadow__[name] = cow_val
        return cow_val
      return val
    raise AttributeError(f"Attribute '{name}' not found on {self.__class__.__name__}")

  def __setattr__(self, name, value):
    if name in ('__proto__', '__shadow__'):
      super().__setattr__(name, value)
    elif self.__proto__ is not None:
      self.__shadow__[name] = value
    else:
      super().__setattr__(name, value)

def _clone_helper(obj, init_fn=None, arena=None):
  if arena is None and hasattr(obj, '__arena__'):
    arena = getattr(obj, '__arena__', None)
  clone_obj = obj.clone()
  if arena is not None:
    arena.register(clone_obj)
  if init_fn:
    init_fn(clone_obj)
  return clone_obj


class Direction(IntEnum):
  North = 0
  East = 1
  South = 2
  West = 3


def get_direction_name(dir):
  _match_res_1 = None
  match dir:
    case Direction.North:
      _match_res_1 = "North"
    case Direction.East:
      _match_res_1 = "East"
    case Direction.South:
      _match_res_1 = "South"
    case _:
      _match_res_1 = "West"
  name = _match_res_1
  return name

def handle_action(code):
  _match_res_2 = None
  match code:
    case 200:
      _match_res_2 = "OK"
    case 404:
      _match_res_2 = "Not Found"
    case _:
      log_msg = "Unknown status code"
      _match_res_2 = log_msg
  result = _match_res_2
  return result

def run_side_effects(dir):
  match dir:
    case Direction.North:
      step = 1
    case _:
      step = 0

def main():
  dir_name = get_direction_name(Direction.South)
  status_str = handle_action(404)
  run_side_effects(Direction.East)
  if ((dir_name == "South") and (status_str == "Not Found")):
    return 0
  return 1

if __name__ == "__main__":
  
  main()
  main()
